packages/core/contracts/financial-templates/optimistic-rewarder/OptimisticRewarder.sol

```diff
@@ -0,0 +1,429 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import "../../common/implementation/Lockable.sol";
+import "../../common/implementation/Testable.sol";
+import "../../oracle/interfaces/FinderInterface.sol";
+import "../../oracle/interfaces/StoreInterface.sol";
+import "../../oracle/interfaces/IdentifierWhitelistInterface.sol";
+import "../../oracle/implementation/Constants.sol";
+import "../../oracle/interfaces/SkinnyOptimisticOracleInterface.sol";
+import "../../common/implementation/AncillaryData.sol";
+import "../../common/interfaces/AddressWhitelistInterface.sol";
+import "./OptimisticRewarderToken.sol";
+
+/**
+ * @notice The base rewarder contract. This manages depositing rewards and paying them out to tokenholders using values
+ * backed by the OptimisticOracle's dispute process.
+ */
+abstract contract OptimisticRewarderBase is Lockable {
+    using SafeERC20 for IERC20;
+
+    struct RedemptionAmount {
+        uint256 amount;
+        IERC20 token;
+    }
+
+    struct Redemption {
+        uint256 finalFee;
+        uint256 expiryTime;
+    }
+
+    // Constants.
+    FinderInterface public finder;
+    uint256 public liveness;
+    uint256 public bond;
+    IERC20 public bondToken;
+    bytes32 public identifier;
+    bytes public customAncillaryData;
+
+    // Mapping to track redemptions.
+    mapping(bytes32 => Redemption) public redemptions;
+
+    // Mapping to track the past total cumulative redemptions for tokenIds.
+    mapping(uint256 => mapping(IERC20 => uint256)) public redeemedAmounts;
```

Comment:
can you explain what the logic is to adding the `redeemedAmounts` to the Base contract rather than adding it to the NFT itself? you could append a prop to each NFT that tracks the cumulative redemption for that token, no?

Reply:
If we depended on this interface or data existing on the NFT, that would mean that other NFTs that were not produced by us and don't implement our custom interfaces couldn't be used in our redemption system. That was the primary reason.